Pagination and Sorting for Get Snapshots API

original-brownbear opened the pull request:
Pagination and snapshots for get snapshots API, build on top of the current implementation to enable work that needs this API for testing. A follow-up will leverage the changes to make things more efficient via pagination.

Relates https://github.com/elastic/elasticsearch/pull/73570 which does part of the under-the-hood changes required to efficiently implement this API on the repository layer.

original-brownbear committed:
bck

original-brownbear committed:
sorting

original-brownbear committed:
bck

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into paginated-get-snapshots-api

original-brownbear committed:
works

original-brownbear committed:
limit

original-brownbear committed:
bck

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into paginated-get-snapshots-api

original-brownbear committed:
bck

original-brownbear committed:
works

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into paginated-get-snapshots-api

original-brownbear committed:
works

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into paginated-get-snapshots-api

original-brownbear committed:
bck

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into paginated-get-snapshots-api

original-brownbear committed:
bck

original-brownbear committed:
better test

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into paginated-get-snapshots-api

original-brownbear committed:
works

original-brownbear committed:
docs ...

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into paginated-get-snapshots-api

original-brownbear committed:
moar coverage

original-brownbear commented on docs/reference/snapshot-restore/apis/get-snapshot-api.asciidoc:
```diff
@@ -98,6 +98,36 @@ comprising the number of shards in the index, the total size of the index in
 bytes, and the maximum number of segments per shard in the index. Defaults to
 `false`, meaning that this information is omitted.
 
+`sort`::
```
The main question left from my end here is on what to do about `?verbose=false` requests. For these, the sorting as implemented here will not work out nicely if the request hits and old-version snapshot that doesn't yet track the timestamps in the `RepositoryData` but will instead just sort by name in this case.

My thinking here would be to maybe just not allow pagination and sorting with `?verbose=false` at all for now. Or alternatively, fail these requests if old version snapshots are found in the repo. 
Another option would be falling back to loading the missing information needed for pagination when running `?verbose=false` queries to work around old repositories but I'm not sure this is worth the effort compared to the other two options.

original-brownbear replied:
> This would only apply if you specify sort, search or size then?

Yes I think just not allowing any of the three for `?verbose=false` would be the plan for option 1.

original-brownbear replied:
Went with option 1 in this PR now => no pagination allowed with `verbose=false`

original-brownbear commented on docs/reference/snapshot-restore/apis/get-snapshot-api.asciidoc:
```diff
@@ -98,6 +98,36 @@ comprising the number of shards in the index, the total size of the index in
+  Sort snapshots by the number of indices they contain and break ties by snapshot name.
+====
+
+`size`::
```
Also, we talked about this before, we want an upper bound here probably in some implicit + explicit form. I think from a response size perspective I'd go with something like `200` for the upper limit on the size in case it's explicitly set and also return a `400` in case a request that does not have `size` set would return more than a certain limit number of snapshots. The only issue with this is that currently there are users for whom something like `3k` snapshots barely works and having an implicit limit would break their use case potentially.

original-brownbear replied:
> Could we leave this untouched if none of sort, search, size are specified at least in 7.x?

I wonder if we can/should :) Obviously we don't want to break the API here, but we are moving towards creating situations where users will have a lot more snapshots that before. We did have problems with the memory consumption of the unpaginated API before and although it's better now, I think we are creating situations where a call getting all snapshot verbose could OOM/destabilize a master easily (thinking about shard failures in particular here). If we do keep the unlimited option untouched in 7.x (which makes sense don't get me wrong :)), should we then maybe invest some effort into some sort of circuit breaking during collecting the `SnapshotInfo` instances to prevent adding a widely used API that can OOM master easily (we could estimate memory from the amount of bytes we read from the blob store or so)?

original-brownbear committed:
nicer formatting

henningandersen commented on docs/reference/snapshot-restore/apis/get-snapshot-api.asciidoc:
```diff
@@ -98,6 +98,36 @@ comprising the number of shards in the index, the total size of the index in
+`after`::
+(Optional, string)
+  Offset from which to start the snapshot listing defined by sort value and snapshot name as a string of the form
+  `<sort_value>,<snapshot_name>`. Defaults to not set which means that listing will start from the first snapshot that matches the
```
I wonder if we should introduce the `next` field into the response too, such that users can just copy that to get the next page? Can certainly also go in a follow-up, I find it less important to have initially at least.

original-brownbear replied:
++ that sounds like a neat idea. Follow-up for that is probably fine, this one is already pretty large :)

DaveCTurner replied:
Hmm `next` sounds more useful than this `after` parameter. Why are we preferring `after`?

original-brownbear replied:
My understanding was that we would return a `next` field and use it as the `after` parameter? So it would just be a usability enhancement. to this parameter?

DaveCTurner replied:
Ah ok. I would rather that the parameter were more opaque then - if we document how to construct it then folk will do just that and then we'll be stuck supporting it even if, say, we want to support multiple sort keys or some other enhancement later on.

original-brownbear replied:
Fine with me, shall we just drop the documentation for the parameter in this PR then and document `next` and `after` in a follow-up?

DaveCTurner replied:
Sure, we could even drop this whole parameter until the followup: IMO `LIMIT` has value even without `OFFSET`. Your call, I'm happy either way.

original-brownbear commented on docs/reference/snapshot-restore/apis/get-snapshot-api.asciidoc:
```diff
@@ -98,6 +98,36 @@ comprising the number of shards in the index, the total size of the index in
+  Offset from which to start the snapshot listing defined by sort value and snapshot name as a string of the form
+  `<sort_value>,<snapshot_name>`. Defaults to not set which means that listing will start from the first snapshot that matches the
+  request.
+
```
I'd prefer this goes into a separate PR due to this size of this one as it adds yet another code path to the get snapshots action I think. That should be a quick one though fortunately :)

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into paginated-get-snapshots-api

original-brownbear committed:
docs update

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into paginated-get-snapshots-api

original-brownbear committed:
cleanup

original-brownbear committed:
spotless

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into paginated-get-snapshots-api

elasticmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

henningandersen reviewed (commented):
Thanks Armin, this looks good. I have not been through the test code yet, but wanted to relay following comments in my first round.

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into paginated-get-snapshots-api

original-brownbear commented on docs/reference/snapshot-restore/apis/get-snapshot-api.asciidoc:
```diff
@@ -98,6 +98,38 @@ comprising the number of shards in the index, the total size of the index in
+`name`::
+  Sort snapshots by their name.
+
+`indices`::
```
This one is effectively free, the necessary information is available via the `RepositoryData` in all versions and the additional code it generates isn't so bad either => I figured I'd do it here?

original-brownbear committed:
fix exception

original-brownbear commented on server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/GetSnapshotsRequest.java:
```diff
@@ -93,6 +112,13 @@ public void writeTo(StreamOutput out) throws IOException {
+            out.writeEnum(sort);
+            out.writeVInt(size);
+        } else if (sort != SortBy.START_TIME || size != 0 || after != null) {
+            throw new IllegalStateException("can't use paginated get snapshots request with node version [" + out.getVersion() + "]");
```
Hmm that's quite tricky to get right to the point of adding that assertion I think. Especially after backporting, we still have to account for the transport client. Also note that we did the same a few lines above for the multi-repo functionality on this request. 

That said you're right, the rest status should come out as `400` here and it should by just using `IllegalArgumentException`. Not sure why I used `IllegalStateException` must have been a typo, sorry for the noise.

original-brownbear commented on server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/GetSnapshotsRequest.java:
```diff
@@ -31,8 +35,18 @@
 
+    public static final Version PAGINATED_GET_SNAPSHOTS_VERSION = Version.V_8_0_0;
+
+    private int size = 0;
```
`-1` is fine by me as well :) My reasoning for `0` was precisely that I found it confusing to forbid `0` and also that it caused more code to have `0` but now that I read it again I think I agree. `-1` is better optics and more consistent :) Will adjust

original-brownbear commented on server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestGetSnapshotsAction.java:
```diff
@@ -53,6 +53,19 @@ public RestChannelConsumer prepareRequest(final RestRequest request, final NodeC
         getSnapshotsRequest.verbose(request.paramAsBoolean("verbose", getSnapshotsRequest.verbose()));
+        final GetSnapshotsRequest.SortBy sort = GetSnapshotsRequest.SortBy.of(request.param("sort", getSnapshotsRequest.sort().toString()));
+        final int size = request.paramAsInt("size", getSnapshotsRequest.size());
+        final String[] afterString = request.paramAsStringArray("after", Strings.EMPTY_ARRAY);
```
Fortunately not :) We validate snapshot names for:

```
        if (snapshotName.contains(",")) {
            throw new InvalidSnapshotNameException(repositoryName, snapshotName, "must not contain ','");
        }
```

We can't have `,` in names because we have APIs like this one that allow fetching snapshots by names comma separated :)

original-brownbear committed:
-1 and proper ex. type

original-brownbear commented on server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java:
```diff
@@ -439,7 +470,101 @@ private boolean isCurrentSnapshotsOnly(String[] snapshots) {
+        return List.copyOf(size > 0 && size < afterStart.size() ? afterStart.subList(0, size) : afterStart);
+    }
+
+    private static boolean nameIsAfter(String name, SnapshotInfo info) {
```
done :)

original-brownbear committed:
multiple params

original-brownbear commented on server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/GetSnapshotsRequest.java:
```diff
@@ -174,6 +211,29 @@ public GetSnapshotsRequest verbose(boolean verbose) {
         return this;
     }
 
+    public GetSnapshotsRequest pagination(@Nullable After after, SortBy sort, int size) {
```
Yea makes sense in hindsight. I had bigger plans here :D but stopped half-way. Moved to separate fields now :) Sorry for the noise

original-brownbear committed:
fix

DaveCTurner commented on docs/reference/snapshot-restore/apis/get-snapshot-api.asciidoc:
```diff
@@ -98,6 +98,38 @@ comprising the number of shards in the index, the total size of the index in
+`name`::
+  Sort snapshots by their name.
+
+`indices`::
```
Maybe `index_count` would be a better name than `indices`? Not that sorting by some amalgamation of the index names is going to be terribly useful, but that might be what people expect with the name `indices`.

original-brownbear replied:
++ renamed

DaveCTurner commented on docs/reference/snapshot-restore/apis/get-snapshot-api.asciidoc:
```diff
@@ -98,6 +98,38 @@ comprising the number of shards in the index, the total size of the index in
 
+`sort`::
+(Optional, string)
+Allows setting a sort order for the result. Defaults to `start_time`, i.e. sorting by snapshot start time stamp.
```
Does this sort them ascending or descending? I'd expect ascending from these docs but in fact if using `start_time` (and probably `duration`) descending would be better. Would it make sense to have a facility to reverse the sort order too?

original-brownbear replied:
> Would it make sense to have a facility to reverse the sort order too?

++ I'd vote yes here, seems useful for the UI in any case. Any objections to adding that @henningandersen ? :)

henningandersen replied:
Agree to adding this.

Reading this documentation line again, I wonder if we should leave the default sorting as "unspecified"? Would we not otherwise risk provoking an unnecessary read of snapshot info for repos that have not been upgraded when using verbose=false?

original-brownbear replied:
I figured this was covered by the fact that the parameter can only be used with verbose listing. Maybe just explicitly call out that verbose=false sort order is unspecified?

DaveCTurner commented on test/framework/src/main/java/org/elasticsearch/snapshots/AbstractSnapshotIntegTestCase.java:
```diff
@@ -531,6 +546,13 @@ public void onTimeout(TimeValue timeout) {
+        return startFullSnapshot(logger, repoName, snapshotName, partial);
+    }
+
+    public static ActionFuture<CreateSnapshotResponse> startFullSnapshot(Logger logger,
```
I don't really like passing a logger around like this, I find it much easier to track down messages if the logger is just the class that logs the message even if it's a base class.

DaveCTurner replied:
It's more a preference than a requirement. Me, I'd have a static logger, I've never had a problem working out which test is running at any given time in the logs, but if you've a good reason for it then that's ok with me.

henningandersen replied:
Just want to weigh in that I also have a preference (but not a requirement) for static loggers.

DaveCTurner reviewed (commented):
Thanks Armin, I left a few thoughts too.

original-brownbear committed:
rename param

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into paginated-get-snapshots-api

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into paginated-get-snapshots-api

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into paginated-get-snapshots-api

original-brownbear committed:
bck

original-brownbear committed:
bck

original-brownbear committed:
bck

original-brownbear committed:
sort order

original-brownbear committed:
works

original-brownbear committed:
ordering works

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into paginated-get-snapshots-api

original-brownbear commented on server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java:
```diff
@@ -403,7 +443,148 @@ private boolean isCurrentSnapshotsOnly(String[] snapshots) {
+
+    private static final Comparator<SnapshotInfo> BY_NAME = Comparator.comparing(sni -> sni.snapshotId().getName());
+
+    private static List<SnapshotInfo> sortSnapshots(
```
This whole method pretty wet now. I opted to keep it like this and not dry it up in some tricky way because it's only here as a place holder for a proper solution anyway, so keeping it simple seemed good enough for now.

original-brownbear commented:
Thanks Henning + David! I addressed all points now I think and added the possibility to specify sort order as well. This should be good for another round :)

henningandersen commented on server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java:
```diff
@@ -403,7 +443,148 @@ private boolean isCurrentSnapshotsOnly(String[] snapshots) {
+        int startIndex = 0;
+        if (after != null) {
+            final String name = after.snapshotName();
+            switch (sortBy) {
```
Can we reuse the comparator to avoid the switching and repetition here? I suppose this is what you meant by your comment above, but unless I am mistaken this would be easy to do?

original-brownbear replied:
Not really right? We don't have a `SnapshotInfo` instance to compare against so we can't use the comparator directly. I refactored the logic here a little now to be drier and use streams to work around the list mutability situation as well, let me know what you think :)

henningandersen commented on server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java:
```diff
@@ -403,7 +443,148 @@ private boolean isCurrentSnapshotsOnly(String[] snapshots) {
+    private static final Comparator<SnapshotInfo> BY_NAME = Comparator.comparing(sni -> sni.snapshotId().getName());
+
+    private static List<SnapshotInfo> sortSnapshots(
+        List<SnapshotInfo> snapshotInfos,
```
This list is mutated by the method, but is also derived from `stream()...collection(Collections.toList())` which does not guarantee that the result is mutable.

Given that this method returns the new list, I would prefer to just do a shallow copy in here.

original-brownbear replied:
solved this via the refactoring to streams now that makes it irrelevant what kind of list we get here

henningandersen commented on docs/reference/snapshot-restore/apis/get-snapshot-api.asciidoc:
```diff
@@ -98,6 +98,37 @@ comprising the number of shards in the index, the total size of the index in
 bytes, and the maximum number of segments per shard in the index. Defaults to
 `false`, meaning that this information is omitted.
 
```
The new arguments are indicated as specified in the body, but I think they have to be specified as a query param?

In fact, looking at some of the other params here, it looks as though those also need to be specified as query param, at least `ignore_unavailable`? I feel like I am missing something that our REST framework handles automatically?

original-brownbear replied:
Hmmm, this seems to be a doc bug, sorry for not noticing. The docs are simply wrong here IMO, all of these parameters only work as parts of the URL. We never parse the request body.
I'll adjust the docs accordingly

original-brownbear replied:
One second look, we have the same bug in at least the snapshot status API as well where we document all query parameters as body elements but don't actually parse the body. Shall we fix this in a separate PR and audit other docs for the same mistake and leave it as is here maybe?

henningandersen commented on qa/smoke-test-http/src/test/java/org/elasticsearch/http/snapshots/RestGetSnapshotsIT.java:
```diff
@@ -0,0 +1,254 @@
+
+        final int inProgressCount = randomIntBetween(6, 20);
+        final List<ActionFuture<CreateSnapshotResponse>> inProgressSnapshots = new ArrayList<>(inProgressCount);
+        AbstractSnapshotIntegTestCase.blockAllDataNodes(repoName);
```
Does this work in a rest test? I think the server runs in a separate jvm there.

original-brownbear replied:
Not in the http smoke tests :) That's why I put this into these.

henningandersen commented on qa/smoke-test-http/src/test/java/org/elasticsearch/http/snapshots/RestGetSnapshotsIT.java:
```diff
@@ -0,0 +1,254 @@
+import static org.hamcrest.Matchers.in;
+import static org.hamcrest.Matchers.is;
+
+// TODO: dry up duplication across this suite and org.elasticsearch.snapshots.GetSnapshotsIT more
```
Yeah, that would be nice. Can be done in a follow-up. I would be OK to have a more minimal REST style test and keep the main sort, pagination and stability testing as an internal cluster IT.

original-brownbear replied:
I basically just wanted to make extra sure all the param parsing is correct and it seemed easiest to test this by simply running a few cases. My reasoning for this one was that we probably won't save that much code but will lose coverage if we do a simpler rest test that still covers all possible param combinations.

henningandersen reviewed (commented):
Some more comments, otherwise looking good.

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into paginated-get-snapshots-api

original-brownbear committed:
drier sorting

original-brownbear commented on docs/reference/snapshot-restore/apis/get-snapshot-api.asciidoc:
```diff
@@ -98,6 +98,37 @@ comprising the number of shards in the index, the total size of the index in
+Sort order. Valid values are `asc` for ascending and `desc` for descending order. Defaults to `asc`, meaning ascending order.
+
+NOTE: The pagination parameters `size`, `order`, and `sort` are not supported when using `verbose=false` and the sort order for
+requests with `verbose=false` is undefined.
```
Currently we're only sorting by name for non-verbose (we sort these thin `SnapshotInfo` that we build in `org.elasticsearch.action.admin.cluster.snapshots.get.TransportGetSnapshotsAction#buildSimpleSnapshotInfos`). I liked undefined better than starting to explain that when we never actually documented a defined sort order for non-verbose. This change for now won't change anything about that APIs return, just a docs adjustment so not really a breaking change? :)

original-brownbear committed:
validation test

original-brownbear commented on server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/GetSnapshotsRequest.java:
```diff
@@ -101,6 +136,20 @@ public ActionRequestValidationException validate() {
+            }
+            if (after != null) {
+                validationException = addValidationError("can't use after with verbose=false", validationException);
+            }
```
++ added that to the validation as well + added a UT for all the validation branches

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into paginated-get-snapshots-api

original-brownbear commented:
Thanks @henningandersen addressed all points I think, this should be ready for another round

henningandersen commented on server/src/test/java/org/elasticsearch/action/admin/cluster/snapshots/get/GetSnapshotsRequestTests.java:
```diff
@@ -0,0 +1,49 @@
+            final ActionRequestValidationException e = request.validate();
+            assertThat(e.getMessage(), containsString("can't use non-default sort order with verbose=false"));
+        }
+        request.after(new GetSnapshotsRequest.After("foo", "bar"));
```
nit: We are accumulating errors here, I wonder if we should either do them one by one or just once with all the errors. I prefer one by one, i.e., a new request object for every validation.

original-brownbear replied:
++ this was kind of lazy :) I adjusted to testing multiple requests.

henningandersen reviewed (approved):
LGTM, thanks for the extra iterations Armin.

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into paginated-get-snapshots-api

original-brownbear committed:
test nits

original-brownbear commented:
Thanks Henning + David! Merging here and moving on to implementing `next` and the search field now.
